slides/refactoring.qmd

```diff
@@ -30,7 +42,92 @@ def print_fastest_time(time_list):
 
 {{< include slides/mentimeter.qmd >}}
 
-## [Exercise: don't repeat yourself (DRY)](https://github.com/{{< meta exercises.org-repo >}}/issues/4)
+## General tips for writing functions
+- small
+- do one thing
+- one level of abstraction per function
+- use descriptive names
+- limit amount of arguments
+
+## How many code smells can we find here?
+
+```{.python}
+a = 500 
+b = 0.75 
+
+# from michael old analysis pipeline
+def process_data(T, sr, t):
+    s = [x * t for i, x in enumerate(T.iloc[:, 0]) if (x % t) > sr and i % 3 == 0]
+    return sorted(s, reverse=True)[:len(s)//2] if len(s) > 5 else s
+
+import pandas as pd
+data = pd.read_csv("/Users/helen/Desktop/29104629.csv")
+
+spikes = process_data(data, a, b)
+```
+
+## How many code smells can we find here?
+
+{{< include slides/mentimeter.qmd >}}
+
+## After refactoring
+```{.python}
+# Constants for sampling rate and spike threshold
+SAMPLING_RATE = 500  # Hz
+SPIKE_THRESHOLD = 0.75  # Arbitrary units
+
+import pandas as pd
+
+# Load the dataset
+# Avoid hardcoded paths and non-meaningful filenames; use a descriptive path.
+data_path = input()
+data = pd.read_csv(data_path)
 
+# Refactored spike detection function with clear logic and meaningful names
+def detect_spikes(data, sampling_rate, threshold):
+    spike_times = []
+    for i, value in enumerate(data.iloc[:, 0]):
+        # Detect spikes based on threshold and other criteria
+        if value > threshold and i % 3 == 0:
+            spike_times.append(value)
+
+    # Return the list of detected spikes, sorted in descending order
+    return sorted(spike_times, reverse=True)
+
+# Detect spikes using the provided sampling rate and threshold
+spikes = detect_spikes(data, SAMPLING_RATE, SPIKE_THRESHOLD)
+```
+
+## Make your code modular
+- preprocess.py
+- analysis.py
+- plot.py
+- io.py (for reading/writing data)
+
+## An example pipeline
+Import your functions and classes in your main analysis pipeline
+```{.python}
+from my_project.io import load_calcium_data
```

Comment:
For this, the students need `pip install -e .[dev]` or similar, right?
But this is only introduced after? How should we deal with this?

Reply:
I think
* either move this later
* use relative imports here, and use the fact that they are not great in some situations to motivate packaging.